scripts/generate_grapheme_properties.cr

```diff
@@ -0,0 +1,79 @@
+# This script generates the file src/string/grapheme/properties.cr
+# that contains compact representations of the GraphemeBreakProperty.txt and emoji-data.txt
+# file from the unicode specification.
+
+require "http/client"
+require "ecr"
+require "compiler/crystal/formatter"
+
+record RRange, low : Int32, high : Int32, prop : String
+
+UCD_ROOT = "http://www.unicode.org/Public/"
+
+def shapeup(arr)
+  i = 0
+  to_del = Array(Int32).new
+  while i < arr.size - 1
+    if arr[i].high + 1 == arr[i + 1].low
+      low = arr[i].low
+      to_del << i
+      arr.delete_at(i)
+      arr[i] = RRange.new(low, arr[i].high, arr[i].prop)
+      i -= 1
+    end
+    i += 1
+  end
+  arr
+end
+
+def parse_graphemes_data(body)
+  result = Hash(String, Array(RRange)).new
+  body.each_line do |line|
+    next unless line = line.strip.presence
+    next if line.starts_with?('#')
+    parts = line.split(';')
+    next unless parts.size >= 2
+
+    fields = parts.first.strip.split("..")
+    f1 = fields.first.to_i(16)
+    f2 = fields.size > 1 ? fields[1].to_i(16) : f1
+
+    prop = parts[1].split('#').first.strip.gsub('_', "")
+
+    (result[prop] ||= Array(RRange).new) << RRange.new(f1, f2, prop)
+  end
+  result.transform_values { |v| shapeup(v) }
+end
+
+def parse_emoji(body)
+  emoji = Array(RRange).new
+  extended = false
+  body.each_line do |line|
+    next unless line = line.strip.presence
+    unless extended
+      extended = line.ends_with?("Extended_Pictographic ; No")
+      next unless extended
+    end
+    next if line.starts_with?('#')
+
+    data = line.split.first.split(';')
+    fields = data.first.split("..")
+    f1 = fields.first.to_i(16)
+    f2 = fields.size > 1 ? fields[1].to_i(16) : f1
+    next if f2 < 0xFF
+    emoji << RRange.new(f1, f2, "ExtendedPictographic")
+  end
+  shapeup(emoji)
+end
+
+body = HTTP::Client.get("#{UCD_ROOT}UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt").body
+props = parse_graphemes_data(body)
+
+body = HTTP::Client.get("#{UCD_ROOT}13.0.0/ucd/emoji/emoji-data.txt").body
```

Comment:
Is it possible not to hardcode a specific version of Unicode here?

Reply:
Would it make sense to integrate it into the existing unicode data script? Then there's just one script for both contexts.